fix(card-editor): allow initial data items

JoelArmendariz opened the pull request:
Closes #1742

**Summary**

Handles initial dataItems, and adds a couple optional props.

**Change List (commits, features, bugs, etc)**

Add:
`getValidTimeRanges`
`onCardChange`

Updates proptypes for `dataItems`

**Acceptance Test (how to verify the PR)**

Go to the `initialValue` and `with custon onCardChange` story in `DashboardEditor` and make sure that the dataseries section in `TimeSeriesCards` works properly

bjornalm commented on src/components/DashboardEditor/DashboardEditor.jsx:
```diff
@@ -307,11 +323,16 @@ const DashboardEditor = ({
                 ...dashboardJson,
                 cards: dashboardJson.cards.map((card) =>
-                  card.id === cardConfig.id ? cardConfig : card
+                  card.id === cardConfig.id
```
Perhaps the `onChange` could be a function with an explanatory name now since the inline function has grown larger

JoelArmendariz replied:
Good call, I extracted it to `handleOnCardChange`

bjornalm reviewed (commented):
Looks good, only have a few suggested changes related to DRYness and readability.

davidicus commented on src/components/CardEditor/CardEditForm/CardEditForm.jsx:
```diff
@@ -47,10 +47,20 @@ const propTypes = {
-  dataItems: PropTypes.arrayOf(PropTypes.string),
+  dataItems: PropTypes.arrayOf(
+    PropTypes.shape({
+      dataSourceId: PropTypes.string,
```
It looks like we are using these values in the component logic but we do not have this as required or do any checks for its value. Is this correct? if so we should do one or the other.

JoelArmendariz replied:
Actually any logic involving the `dataItems` acts properly because it defaults to an empty array when not present. I was unable to find anywhere that it would cause problems.

bjornalm commented:
I don't find the story `with custon onCardChange`.

When testing story `with initial value` I found this:
a)
1. Go to https://deploy-preview-1743--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-experimental-dashboardeditor--with-initial-value
2. Click Time series line in the gallary
3. Now click the existing TimeSeries card in the dashboard grid
4. Notice how the Data item in the inspector panel is not showing anything, unlike if yo would have clicked the TimeSeries card in the dashboard grid directly

b)
1. Go to https://deploy-preview-1743--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-experimental-dashboardeditor--with-initial-value
2. Click the existing TimeSeries card in the dashboard grid
3. Uncheck the selection in Data item dropdown
4. Notice how the Data item dropdown is empty
![image](https://user-images.githubusercontent.com/5167091/98124329-55a83b80-1eb3-11eb-9754-22e0ba5e000e.png)

JoelArmendariz commented on src/components/CardEditor/CardEditForm/CardEditFormItems/DataSeriesFormItem.jsx:
```diff
@@ -214,11 +225,12 @@ const DataSeriesFormItem = ({
-              ? validDataItems.map((dataItem) => ({
-                  id: dataItem,
-                  text: dataItem,
+              ? validDataItems.map(({ dataSourceId, label }) => ({
```
I have changed this to a reuse-able function.

JoelArmendariz commented:
> I don't find the story `with custon onCardChange`.
> 
> When testing story `with initial value` I found this:
> a)
> 
>     1. Go to https://deploy-preview-1743--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-experimental-dashboardeditor--with-initial-value
> 
>     2. Click Time series line in the gallary
> 
>     3. Now click the existing TimeSeries card in the dashboard grid
> 
>     4. Notice how the Data item in the inspector panel is not showing anything, unlike if yo would have clicked the TimeSeries card in the dashboard grid directly
> 
> 
> b)
> 
>     1. Go to https://deploy-preview-1743--carbon-addons-iot-react.netlify.app/?path=/story/watson-iot-experimental-dashboardeditor--with-initial-value
> 
>     2. Click the existing TimeSeries card in the dashboard grid
> 
>     3. Uncheck the selection in Data item dropdown
> 
>     4. Notice how the Data item dropdown is empty
>        ![image](https://user-images.githubusercontent.com/5167091/98124329-55a83b80-1eb3-11eb-9754-22e0ba5e000e.png)

Ah weird.. The merge got rid of that story, sorry about that. I added it back.
I also fixed the empty dropdown problem. @bjornalm

JoelArmendariz commented on src/components/CardEditor/CardEditForm/CardEditFormItems/DataSeriesFormItem.jsx:
```diff
@@ -61,10 +61,15 @@ const propTypes = {
    * this prop will be ignored if getValidDataItems is defined
    */
-  dataItems: PropTypes.arrayOf(PropTypes.string),
+  dataItems: PropTypes.arrayOf(
```
I have extracted this proptype to be re-used after one definition.

bjornalm commented on src/components/CardEditor/CardEditForm/CardEditFormItems/DataSeriesFormItem.jsx:
```diff
@@ -125,6 +125,12 @@ export const formatSeries = (selectedItems, cardJson) => {
   return series;
 };
 
+export const formatDataItemsForDropdown = (dataItems) =>
```
Does it have to be exported?

JoelArmendariz replied:
Oh my intentions were to add a test case, but I forgot. Will add into the next iteration

bjornalm commented:
Nice work!

tay1orjones commented:
:tada: This PR is included in version 2.140.1-next.1 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/carbon-addons-iot-react/v/2.140.1-next.1)
- [GitHub release](https://github.com/carbon-design-system/carbon-addons-iot-react/releases/tag/v2.140.1-next.1)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

tay1orjones commented:
:tada: This PR is included in version 2.141.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/carbon-addons-iot-react/v/2.141.0)
- [GitHub release](https://github.com/carbon-design-system/carbon-addons-iot-react/releases/tag/v2.141.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
